bin/ch/Helpers.cpp

```diff
@@ -42,81 +41,65 @@ HRESULT Helpers::LoadScriptFromFile(LPCSTR filename, LPCSTR& contents, bool* isU
         IfFailGo(E_FAIL);
     }
 
-    //
-    // Determine the file length, in bytes.
-    //
-    fseek(file, 0, SEEK_END);
-    lengthBytes = ftell(file);
-    fseek(file, 0, SEEK_SET);
-    pRawBytes = (LPBYTE)malloc(lengthBytes + sizeof(WCHAR));
-    if (nullptr == pRawBytes)
-    {
-        fwprintf(stderr, _u("out of memory"));
-        IfFailGo(E_OUTOFMEMORY);
-    }
-
-    //
-    // Read the entire content as a binary block.
-    //
-    fread(pRawBytes, sizeof(BYTE), lengthBytes, file);
-    fclose(file);
-    *reinterpret_cast<WCHAR*>(pRawBytes + lengthBytes) = 0; // Null terminate it. Could be UTF16
-
-    //
-    // Read encoding, handling any conversion to Unicode.
-    //
-    // Warning: The UNICODE buffer for parsing is supposed to be provided by the host.
-    // This is not a complete read of the encoding. Some encodings like UTF7, UTF1, EBCDIC, SCSU, BOCU could be
-    // wrongly classified as ANSI
-    //
+    if (file != nullptr)
     {
-        LPCWSTR contentsRaw = reinterpret_cast<LPCWSTR>(pRawBytes);
-        if ((0xEF == *pRawBytes && 0xBB == *(pRawBytes + 1) && 0xBF == *(pRawBytes + 2)))
-        {
-            isUtf8 = true;
-        }
-        else if (0xFFFE == *contentsRaw || (0x0000 == *contentsRaw && 0xFEFF == *(contentsRaw + 1)))
-        {
-            // unicode unsupported
-            fwprintf(stderr, _u("unsupported file encoding"));
-            IfFailGo(E_UNEXPECTED);
-        }
-        else if (0xFEFF == *contentsRaw)
+        // Determine the file length, in bytes.
+        fseek(file, 0, SEEK_END);
+        lengthBytes = ftell(file);
+        fseek(file, 0, SEEK_SET);
+        pRawBytes = (LPBYTE)malloc(lengthBytes + sizeof(BYTE));
+        if (nullptr == pRawBytes)
         {
-            // unicode LE
-            isUtf8 = false;
+            fwprintf(stderr, _u("out of memory"));
+            IfFailGo(E_OUTOFMEMORY);
         }
-        else
+
+        //
+        // Read the entire content as a binary block.
+        //
+        size_t readBytes = fread(pRawBytes, sizeof(BYTE), lengthBytes, file);
+        if (readBytes == 0)
         {
-            // Assume UTF8
-            isUtf8 = true;
+            IfFailGo(E_FAIL);
         }
-    }
 
-    if (isUtf8)
-    {
+        *reinterpret_cast<BYTE*>(pRawBytes + lengthBytes) = 0; // Null terminate it. Could be UTF16
+
         contents = reinterpret_cast<LPCSTR>(pRawBytes);
-    }
-    else
-    {
-        LPSTR pNarrow = nullptr;
-        IfFailGo(WideStringToNarrowDynamic(reinterpret_cast<LPCWSTR>(pRawBytes), &pNarrow));
-        contents = pNarrow;
+        //
+        // Read encoding to make sure it's supported
+        //
+        // Warning: The UNICODE buffer for parsing is supposed to be provided by the host.
+        // This is not a complete read of the encoding. Some encodings like UTF7, UTF1, EBCDIC, SCSU, BOCU could be
+        // wrongly classified as ANSI
+        //
+        {
+            C_ASSERT(sizeof(WCHAR) == 2);
```

Comment:
sizeof(char16) is being bigger than 2 does not necessarily mean that we wouldn't support it.

Instead of casting from char to char16 below, you may compare the UTF-16 BOM using char type. At file level, char16 would never be more than 2 bytes. So the exception of char16 is being 32bits or even 64bits is independent from what it actually represents.

IMO converting from char16 to utf-8 is straightforward. Instead of marking it as unsupported, we can also add the conversion without requiring ICU etc.

> Some encodings like UTF7, UTF1, EBCDIC, SCSU, BOCU could be wrongly classified as ANSI

This mean that we support them or not ? We only check for utf-16 BOM below.